{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from typing import List, Tuple, Dict, Any, Optional\n",
    "import pickle\n",
    "\n",
    "os.chdir('/Users/xbh0403/Desktop/TWAS_ASSOC')\n",
    "\n",
    "from sklearn import linear_model\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from pyplink import PyPlink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes of the dataframes: (373, 13942) (89, 13942) (13942, 5) (13942, 5)\n"
     ]
    }
   ],
   "source": [
    "def load_data():\n",
    "    EUR_ge_regressed = pd.read_csv(\"./project_data/GEUVADIS_EUR_ge_regressed.tsv.gz\", sep=\"\\t\", index_col=0, compression=\"gzip\")\n",
    "    YRI_ge_regressed = pd.read_csv(\"./project_data/GEUVADIS_YRI_ge_regressed.tsv.gz\", sep=\"\\t\", index_col=0, compression=\"gzip\")\n",
    "\n",
    "    EUR_protein_genes = pd.read_csv(\"./project_data/GEUVADIS_EUR_protein_genes.tsv.gz\", sep=\"\\t\", index_col=0, compression=\"gzip\")\n",
    "    EUR_protein_genes[\"chr\"] = EUR_protein_genes.index\n",
    "    EUR_protein_genes.reset_index(drop=True, inplace=True)\n",
    "    YRI_protein_genes = pd.read_csv(\"./project_data/GEUVADIS_YRI_protein_genes.tsv.gz\", sep=\"\\t\", index_col=0, compression=\"gzip\")\n",
    "    YRI_protein_genes[\"chr\"] = YRI_protein_genes.index\n",
    "    YRI_protein_genes.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    print(\"Shapes of the dataframes:\", EUR_ge_regressed.shape, YRI_ge_regressed.shape, EUR_protein_genes.shape, YRI_protein_genes.shape)\n",
    "\n",
    "    return EUR_ge_regressed, YRI_ge_regressed, EUR_protein_genes, YRI_protein_genes\n",
    "\n",
    "EUR_ge_regressed, YRI_ge_regressed, EUR_protein_genes, YRI_protein_genes = load_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000187634</th>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <th>ENSG00000187961</th>\n",
       "      <th>ENSG00000187583</th>\n",
       "      <th>ENSG00000187642</th>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <th>ENSG00000187608</th>\n",
       "      <th>ENSG00000188157</th>\n",
       "      <th>ENSG00000131591</th>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000187609</th>\n",
       "      <th>ENSG00000188747</th>\n",
       "      <th>ENSG00000165802</th>\n",
       "      <th>ENSG00000130653</th>\n",
       "      <th>ENSG00000182154</th>\n",
       "      <th>ENSG00000148399</th>\n",
       "      <th>ENSG00000165724</th>\n",
       "      <th>ENSG00000197070</th>\n",
       "      <th>ENSG00000181090</th>\n",
       "      <th>ENSG00000148408</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA18486</th>\n",
       "      <td>0.094744</td>\n",
       "      <td>-0.351085</td>\n",
       "      <td>-1.932980</td>\n",
       "      <td>-0.321105</td>\n",
       "      <td>-0.515498</td>\n",
       "      <td>-2.079754</td>\n",
       "      <td>-0.273891</td>\n",
       "      <td>0.033437</td>\n",
       "      <td>0.364211</td>\n",
       "      <td>-0.428842</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.928062</td>\n",
       "      <td>-1.769592</td>\n",
       "      <td>0.392149</td>\n",
       "      <td>0.521271</td>\n",
       "      <td>0.846738</td>\n",
       "      <td>-0.396933</td>\n",
       "      <td>0.285021</td>\n",
       "      <td>-0.399697</td>\n",
       "      <td>1.098842</td>\n",
       "      <td>-0.650935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA18487</th>\n",
       "      <td>-0.902019</td>\n",
       "      <td>-1.003504</td>\n",
       "      <td>-0.070440</td>\n",
       "      <td>-0.136875</td>\n",
       "      <td>-0.813910</td>\n",
       "      <td>-1.176463</td>\n",
       "      <td>0.851783</td>\n",
       "      <td>-0.342203</td>\n",
       "      <td>-2.013583</td>\n",
       "      <td>-1.726217</td>\n",
       "      <td>...</td>\n",
       "      <td>0.363013</td>\n",
       "      <td>1.649147</td>\n",
       "      <td>0.490006</td>\n",
       "      <td>-1.074849</td>\n",
       "      <td>0.904444</td>\n",
       "      <td>-0.928404</td>\n",
       "      <td>-1.029141</td>\n",
       "      <td>1.378654</td>\n",
       "      <td>1.601134</td>\n",
       "      <td>-1.061971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA18488</th>\n",
       "      <td>0.787695</td>\n",
       "      <td>0.997348</td>\n",
       "      <td>-0.302960</td>\n",
       "      <td>0.569240</td>\n",
       "      <td>0.139434</td>\n",
       "      <td>0.274250</td>\n",
       "      <td>-0.596212</td>\n",
       "      <td>0.165744</td>\n",
       "      <td>0.722004</td>\n",
       "      <td>-1.196606</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.730004</td>\n",
       "      <td>-0.802126</td>\n",
       "      <td>1.158687</td>\n",
       "      <td>1.805478</td>\n",
       "      <td>0.252108</td>\n",
       "      <td>-1.789603</td>\n",
       "      <td>-0.507488</td>\n",
       "      <td>-0.404853</td>\n",
       "      <td>1.079572</td>\n",
       "      <td>-0.565762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA18489</th>\n",
       "      <td>-0.424581</td>\n",
       "      <td>0.991987</td>\n",
       "      <td>-0.191124</td>\n",
       "      <td>0.463024</td>\n",
       "      <td>0.538183</td>\n",
       "      <td>-0.787763</td>\n",
       "      <td>-0.231541</td>\n",
       "      <td>1.012617</td>\n",
       "      <td>0.530263</td>\n",
       "      <td>-0.254174</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.166257</td>\n",
       "      <td>0.819447</td>\n",
       "      <td>1.003430</td>\n",
       "      <td>-0.000927</td>\n",
       "      <td>-0.424602</td>\n",
       "      <td>-0.038372</td>\n",
       "      <td>-0.091295</td>\n",
       "      <td>-1.516224</td>\n",
       "      <td>0.934990</td>\n",
       "      <td>-1.346285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA18498</th>\n",
       "      <td>-0.873080</td>\n",
       "      <td>-0.675177</td>\n",
       "      <td>0.048259</td>\n",
       "      <td>-1.092280</td>\n",
       "      <td>0.573952</td>\n",
       "      <td>0.626174</td>\n",
       "      <td>-0.147917</td>\n",
       "      <td>0.465355</td>\n",
       "      <td>0.619457</td>\n",
       "      <td>-0.494885</td>\n",
       "      <td>...</td>\n",
       "      <td>0.757072</td>\n",
       "      <td>-0.423255</td>\n",
       "      <td>0.246138</td>\n",
       "      <td>-0.073313</td>\n",
       "      <td>0.088300</td>\n",
       "      <td>0.017429</td>\n",
       "      <td>-0.811266</td>\n",
       "      <td>0.273057</td>\n",
       "      <td>2.637008</td>\n",
       "      <td>-0.349992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19236</th>\n",
       "      <td>-0.719438</td>\n",
       "      <td>0.514831</td>\n",
       "      <td>1.334156</td>\n",
       "      <td>-0.265431</td>\n",
       "      <td>1.475308</td>\n",
       "      <td>-1.031606</td>\n",
       "      <td>-0.005093</td>\n",
       "      <td>0.409364</td>\n",
       "      <td>-0.029977</td>\n",
       "      <td>-0.324761</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.779381</td>\n",
       "      <td>-1.241429</td>\n",
       "      <td>0.302309</td>\n",
       "      <td>0.729772</td>\n",
       "      <td>-0.236441</td>\n",
       "      <td>0.554705</td>\n",
       "      <td>-0.371973</td>\n",
       "      <td>0.184894</td>\n",
       "      <td>-0.352576</td>\n",
       "      <td>0.254183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19247</th>\n",
       "      <td>0.488869</td>\n",
       "      <td>0.378698</td>\n",
       "      <td>-0.522324</td>\n",
       "      <td>-0.267890</td>\n",
       "      <td>-0.828062</td>\n",
       "      <td>1.360805</td>\n",
       "      <td>-0.170798</td>\n",
       "      <td>0.100119</td>\n",
       "      <td>-0.029440</td>\n",
       "      <td>-0.231236</td>\n",
       "      <td>...</td>\n",
       "      <td>0.310839</td>\n",
       "      <td>0.378431</td>\n",
       "      <td>-0.850058</td>\n",
       "      <td>-1.266535</td>\n",
       "      <td>1.118367</td>\n",
       "      <td>0.303224</td>\n",
       "      <td>-1.360509</td>\n",
       "      <td>0.627126</td>\n",
       "      <td>-1.371108</td>\n",
       "      <td>0.134620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19248</th>\n",
       "      <td>1.038800</td>\n",
       "      <td>1.203831</td>\n",
       "      <td>0.051683</td>\n",
       "      <td>-0.004533</td>\n",
       "      <td>0.150353</td>\n",
       "      <td>0.020862</td>\n",
       "      <td>0.338120</td>\n",
       "      <td>1.413611</td>\n",
       "      <td>0.678311</td>\n",
       "      <td>0.289597</td>\n",
       "      <td>...</td>\n",
       "      <td>0.629750</td>\n",
       "      <td>-0.256761</td>\n",
       "      <td>-0.026252</td>\n",
       "      <td>-0.836387</td>\n",
       "      <td>-0.844268</td>\n",
       "      <td>0.802872</td>\n",
       "      <td>-0.047072</td>\n",
       "      <td>-0.640519</td>\n",
       "      <td>-0.734871</td>\n",
       "      <td>-0.343510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19256</th>\n",
       "      <td>-1.718986</td>\n",
       "      <td>-0.760087</td>\n",
       "      <td>0.382552</td>\n",
       "      <td>0.990427</td>\n",
       "      <td>0.493457</td>\n",
       "      <td>-0.789862</td>\n",
       "      <td>-0.642850</td>\n",
       "      <td>0.329378</td>\n",
       "      <td>0.879047</td>\n",
       "      <td>-0.565545</td>\n",
       "      <td>...</td>\n",
       "      <td>1.304044</td>\n",
       "      <td>1.385512</td>\n",
       "      <td>0.332867</td>\n",
       "      <td>0.383803</td>\n",
       "      <td>-1.767664</td>\n",
       "      <td>1.280153</td>\n",
       "      <td>-0.669369</td>\n",
       "      <td>0.240813</td>\n",
       "      <td>0.055344</td>\n",
       "      <td>0.389836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA19257</th>\n",
       "      <td>0.718940</td>\n",
       "      <td>0.574684</td>\n",
       "      <td>-1.483772</td>\n",
       "      <td>-0.496506</td>\n",
       "      <td>-1.018358</td>\n",
       "      <td>0.863802</td>\n",
       "      <td>-1.094281</td>\n",
       "      <td>-0.053911</td>\n",
       "      <td>-0.844273</td>\n",
       "      <td>-0.025472</td>\n",
       "      <td>...</td>\n",
       "      <td>0.212884</td>\n",
       "      <td>-0.795647</td>\n",
       "      <td>-0.726404</td>\n",
       "      <td>-0.241439</td>\n",
       "      <td>0.920960</td>\n",
       "      <td>0.829995</td>\n",
       "      <td>0.109434</td>\n",
       "      <td>-0.552956</td>\n",
       "      <td>-1.146548</td>\n",
       "      <td>-1.863217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows Ã— 13942 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ENSG00000187634  ENSG00000188976  ENSG00000187961  ENSG00000187583  \\\n",
       "NA18486         0.094744        -0.351085        -1.932980        -0.321105   \n",
       "NA18487        -0.902019        -1.003504        -0.070440        -0.136875   \n",
       "NA18488         0.787695         0.997348        -0.302960         0.569240   \n",
       "NA18489        -0.424581         0.991987        -0.191124         0.463024   \n",
       "NA18498        -0.873080        -0.675177         0.048259        -1.092280   \n",
       "...                  ...              ...              ...              ...   \n",
       "NA19236        -0.719438         0.514831         1.334156        -0.265431   \n",
       "NA19247         0.488869         0.378698        -0.522324        -0.267890   \n",
       "NA19248         1.038800         1.203831         0.051683        -0.004533   \n",
       "NA19256        -1.718986        -0.760087         0.382552         0.990427   \n",
       "NA19257         0.718940         0.574684        -1.483772        -0.496506   \n",
       "\n",
       "         ENSG00000187642  ENSG00000188290  ENSG00000187608  ENSG00000188157  \\\n",
       "NA18486        -0.515498        -2.079754        -0.273891         0.033437   \n",
       "NA18487        -0.813910        -1.176463         0.851783        -0.342203   \n",
       "NA18488         0.139434         0.274250        -0.596212         0.165744   \n",
       "NA18489         0.538183        -0.787763        -0.231541         1.012617   \n",
       "NA18498         0.573952         0.626174        -0.147917         0.465355   \n",
       "...                  ...              ...              ...              ...   \n",
       "NA19236         1.475308        -1.031606        -0.005093         0.409364   \n",
       "NA19247        -0.828062         1.360805        -0.170798         0.100119   \n",
       "NA19248         0.150353         0.020862         0.338120         1.413611   \n",
       "NA19256         0.493457        -0.789862        -0.642850         0.329378   \n",
       "NA19257        -1.018358         0.863802        -1.094281        -0.053911   \n",
       "\n",
       "         ENSG00000131591  ENSG00000186891  ...  ENSG00000187609  \\\n",
       "NA18486         0.364211        -0.428842  ...        -0.928062   \n",
       "NA18487        -2.013583        -1.726217  ...         0.363013   \n",
       "NA18488         0.722004        -1.196606  ...        -0.730004   \n",
       "NA18489         0.530263        -0.254174  ...        -1.166257   \n",
       "NA18498         0.619457        -0.494885  ...         0.757072   \n",
       "...                  ...              ...  ...              ...   \n",
       "NA19236        -0.029977        -0.324761  ...        -0.779381   \n",
       "NA19247        -0.029440        -0.231236  ...         0.310839   \n",
       "NA19248         0.678311         0.289597  ...         0.629750   \n",
       "NA19256         0.879047        -0.565545  ...         1.304044   \n",
       "NA19257        -0.844273        -0.025472  ...         0.212884   \n",
       "\n",
       "         ENSG00000188747  ENSG00000165802  ENSG00000130653  ENSG00000182154  \\\n",
       "NA18486        -1.769592         0.392149         0.521271         0.846738   \n",
       "NA18487         1.649147         0.490006        -1.074849         0.904444   \n",
       "NA18488        -0.802126         1.158687         1.805478         0.252108   \n",
       "NA18489         0.819447         1.003430        -0.000927        -0.424602   \n",
       "NA18498        -0.423255         0.246138        -0.073313         0.088300   \n",
       "...                  ...              ...              ...              ...   \n",
       "NA19236        -1.241429         0.302309         0.729772        -0.236441   \n",
       "NA19247         0.378431        -0.850058        -1.266535         1.118367   \n",
       "NA19248        -0.256761        -0.026252        -0.836387        -0.844268   \n",
       "NA19256         1.385512         0.332867         0.383803        -1.767664   \n",
       "NA19257        -0.795647        -0.726404        -0.241439         0.920960   \n",
       "\n",
       "         ENSG00000148399  ENSG00000165724  ENSG00000197070  ENSG00000181090  \\\n",
       "NA18486        -0.396933         0.285021        -0.399697         1.098842   \n",
       "NA18487        -0.928404        -1.029141         1.378654         1.601134   \n",
       "NA18488        -1.789603        -0.507488        -0.404853         1.079572   \n",
       "NA18489        -0.038372        -0.091295        -1.516224         0.934990   \n",
       "NA18498         0.017429        -0.811266         0.273057         2.637008   \n",
       "...                  ...              ...              ...              ...   \n",
       "NA19236         0.554705        -0.371973         0.184894        -0.352576   \n",
       "NA19247         0.303224        -1.360509         0.627126        -1.371108   \n",
       "NA19248         0.802872        -0.047072        -0.640519        -0.734871   \n",
       "NA19256         1.280153        -0.669369         0.240813         0.055344   \n",
       "NA19257         0.829995         0.109434        -0.552956        -1.146548   \n",
       "\n",
       "         ENSG00000148408  \n",
       "NA18486        -0.650935  \n",
       "NA18487        -1.061971  \n",
       "NA18488        -0.565762  \n",
       "NA18489        -1.346285  \n",
       "NA18498        -0.349992  \n",
       "...                  ...  \n",
       "NA19236         0.254183  \n",
       "NA19247         0.134620  \n",
       "NA19248        -0.343510  \n",
       "NA19256         0.389836  \n",
       "NA19257        -1.863217  \n",
       "\n",
       "[89 rows x 13942 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "YRI_ge_regressed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_snps_in_gene(chr_num: int, start: int, end: int, ancsetry: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Find SNPs within a specified genomic region.\n",
    "\n",
    "    Parameters:\n",
    "    chr_num (int): Chromosome number.\n",
    "    start (int): Start position of the genomic region.\n",
    "    end (int): End position of the genomic region.\n",
    "    bim (pd.DataFrame): DataFrame containing SNP information.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: DataFrame containing SNPs within the specified genomic region.\n",
    "    \n",
    "    \"\"\"\n",
    "    with PyPlink(\"./project_data/geno/\"+ancsetry+\"/GEUVADIS_\"+ancsetry+\"_chr\"+str(chr_num)) as bed:\n",
    "        # Getting the BIM and FAM\n",
    "        bim = bed.get_bim()\n",
    "        bim[\"snp\"] = bim.index\n",
    "        bim.reset_index(drop=True, inplace=True)\n",
    "        bim = bim[[\"snp\", \"chrom\", \"pos\", \"cm\", \"a1\", \"a2\"]]\n",
    "    return bim[(bim[\"pos\"] >= start) & (bim[\"pos\"] <= end)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_geno(ancsetry: str, chr_num: int, snps: List[str], start: int, gene_id: str, save_result: bool = False) -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Process the genotype data.\n",
    "\n",
    "    Parameters:\n",
    "    ancsetry (str): Ancestry of the individual.\n",
    "    chr_num (int): Chromosome number.\n",
    "    individual (str | None): Individual ID.\n",
    "    snps (List[str]): List of SNPs.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: Processed genotype data.\n",
    "    \"\"\"\n",
    "    if len(snps) == 0:\n",
    "        raise ValueError(\"No SNPs provided.\")\n",
    "    \n",
    "    with PyPlink(\"./project_data/geno/\"+ancsetry+\"/GEUVADIS_\"+ancsetry+\"_chr\"+str(chr_num)) as bed:\n",
    "        # Getting the BIM and FAM\n",
    "        bim = bed.get_bim()\n",
    "        bim[\"snp\"] = bim.index\n",
    "        bim.reset_index(drop=True, inplace=True)\n",
    "        bim = bim[[\"snp\", \"chrom\", \"pos\", \"cm\", \"a1\", \"a2\"]]\n",
    "        snp_info = bim[bim[\"snp\"].isin(snps)]\n",
    "\n",
    "        fam = bed.get_fam()\n",
    "        iids = fam[\"iid\"].tolist()\n",
    "\n",
    "        results = {\n",
    "            \"snp_info\": snp_info,\n",
    "            \"iids\": iids,\n",
    "        }\n",
    "\n",
    "        keep_bool = [1] * len(iids)\n",
    "\n",
    "        for snp_id, genotypes in bed.iter_geno_marker(snps):\n",
    "            results[snp_id] = genotypes\n",
    "            genotypes_kept = [0 if x == -1 else 1 for x in genotypes]\n",
    "            keep_bool = [x*y for x, y in zip(keep_bool, genotypes_kept)]\n",
    "\n",
    "        results[\"keep_bool\"] = keep_bool\n",
    "\n",
    "        for snp_id, genotypes in bed.iter_geno_marker(snps):\n",
    "            results[snp_id+\"_filtered\"] = [x for x, y in zip(genotypes, keep_bool) if y == 1]\n",
    "        \n",
    "        results[\"iids_filtered\"] = [x for x, y in zip(iids, keep_bool) if y == 1]\n",
    "\n",
    "        X = np.zeros((len(results[\"iids_filtered\"]), snp_info.shape[0]))\n",
    "\n",
    "        for i in range(snp_info.shape[0]):\n",
    "            X[:, i] = results[snps[i]+\"_filtered\"]\n",
    "\n",
    "        if save_result:\n",
    "            with open(\"./project_data/processed_Xy/\"+gene_id+\"_results.pkl\", \"wb\") as f:\n",
    "                pickle.dump(results, f)\n",
    "            # save x as npy\n",
    "            np.save(\"./project_data/processed_Xy/X/\"+gene_id+\"_X.npy\", X)\n",
    "\n",
    "        return results, X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_y(gene_id: str, iids: List[str], y_full_df: pd.DataFrame = EUR_ge_regressed, save_y: bool = False) -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Get the Y values for a specific gene and ancestry.\n",
    "\n",
    "    Parameters:\n",
    "    gene_id (str): Gene ID.\n",
    "    iids (List[str]): List of individual IDs.\n",
    "    y_full_df (pd.DataFrame): DataFrame containing the Y values.\n",
    "\n",
    "    Returns:\n",
    "    np.ndarray: Array of Y values for the specified gene and ancestry.\n",
    "    \"\"\"\n",
    "    all_columns = y_full_df.columns\n",
    "    id_gene = all_columns.get_loc(gene_id)\n",
    "    all_individuals = y_full_df.index\n",
    "    iids_keep = [True if x in iids else False for x in all_individuals]\n",
    "    y = y_full_df.iloc[iids_keep, id_gene].values\n",
    "    if save_y:\n",
    "        np.save(\"./project_data/processed_Xy/y/\"+gene_id+\"_y.npy\", y)\n",
    "    return y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_one_gene(gene_id: str, protein_genes: pd.DataFrame, ancsetry: str) -> np.ndarray:\n",
    "    gene = protein_genes[protein_genes[\"gene_id\"] == gene_id]\n",
    "    assert gene.shape[0] == 1\n",
    "    chr_num = gene[\"chr\"].values[0]\n",
    "    start = gene[\"start\"].values[0]\n",
    "    end = gene[\"end\"].values[0]\n",
    "    gene_name = gene[\"name\"].values[0]\n",
    "    print(\"Gene id: \", gene_id, \" Gene name: \", gene_name, \" Chr: \", chr_num, \" Start: \", start, \" End: \", end)\n",
    "    \n",
    "    snps = find_snps_in_gene(chr_num, start, end, ancsetry)\n",
    "    snps_name = snps[\"snp\"].tolist()\n",
    "    print(\"Number of SNPs: \", len(snps_name))\n",
    "    processed_geno, X = process_geno(ancsetry, chr_num, snps_name, start, gene_id, save_result=True)\n",
    "    y = get_y(gene_id, processed_geno[\"iids_filtered\"], save_y=True)\n",
    "    return processed_geno, X, y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene id:  ENSG00000187642  Gene name:  PERM1  Chr:  1  Start:  410579  End:  1410580\n",
      "Number of SNPs:  138\n"
     ]
    }
   ],
   "source": [
    "# one_gene_result, X, y = process_one_gene(\"ENSG00000187634\", EUR_protein_genes, \"EUR\")\n",
    "one_gene_result, X, y = process_one_gene(\"ENSG00000187642\", EUR_protein_genes, \"EUR\")\n",
    "# one_gene_result, X, y = process_one_gene(\"ENSG00000187583\", EUR_protein_genes, \"EUR\")\n",
    "# one_gene_result, X, y = process_one_gene(\"ENSG00000187961\", EUR_protein_genes, \"EUR\")\n",
    "# one_gene_result, X, y = process_one_gene(\"ENSG00000144182\", EUR_protein_genes, \"EUR\")\n",
    "# one_gene_result, X, y = process_one_gene(\"ENSG00000215784\", EUR_protein_genes, \"EUR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def standardize(X: np.ndarray) -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Standardize the X matrix.\n",
    "\n",
    "    Parameters:\n",
    "    X (np.ndarray): X matrix.\n",
    "\n",
    "    Returns:\n",
    "    np.ndarray: Standardized X matrix.\n",
    "    \"\"\"\n",
    "    mean = np.mean(X, axis=0)\n",
    "    std = np.std(X, axis=0)\n",
    "    X = [(X[i, :] - mean) / std for i in range(X.shape[0]) if std[i] != 0]\n",
    "    return X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "X_std = StandardScaler().fit_transform(X)\n",
    "X_train, X_test, y_train, y_test = train_test_split(X_std, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean squared error:  0.5590618915480747\n",
      "Mean squared error:  3.664491328632582e+26\n"
     ]
    }
   ],
   "source": [
    "# Linear regression\n",
    "reg = linear_model.LinearRegression()\n",
    "reg.fit(X_train, y_train)\n",
    "y_pred = reg.predict(X_test)\n",
    "y_pred_train = reg.predict(X_train)\n",
    "print(\"Mean squared error: \", np.mean((y_train - y_pred_train)**2))\n",
    "print(\"Mean squared error: \", np.mean((y_pred - y_test)**2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.03634878314031986\n",
      "-0.022963852380138983\n"
     ]
    }
   ],
   "source": [
    "lasso = linear_model.Lasso(alpha=0.1, random_state=42)\n",
    "lasso.fit(X_train, y_train)\n",
    "print(lasso.score(X_train, y_train))\n",
    "print(lasso.score(X_test, y_test))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def LASSO(X, Y, alpha=0.1, test_size=0.2, random_state=42):\n",
    "    X_std = StandardScaler().fit_transform(X)\n",
    "    X_train, X_test, Y_train, Y_test = train_test_split(X_std, Y, test_size=test_size, random_state=random_state)\n",
    "    clf = linear_model.Lasso(alpha=alpha)\n",
    "    clf.fit(X_train, Y_train)\n",
    "    return {\n",
    "        \"clf\": clf,\n",
    "        \"X_train\": X_train,\n",
    "        \"X_test\": X_test,\n",
    "        \"Y_train\": Y_train,\n",
    "        \"Y_train_pred\": clf.predict(X_train),\n",
    "        \"Y_test\": Y_test,\n",
    "        \"Y_test_pred\": clf.predict(X_test),\n",
    "        \"r2_train\": clf.score(X_train, Y_train),\n",
    "        \"r2_test\": clf.score(X_test, Y_test)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "lasso_result = LASSO(X, y, alpha=0.1, test_size=0.2, random_state=42)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
